File: Arrays_Hashing/TopKFrequentElements/TopKFrequentElements.kt
class Solution {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val map = HashMap<Int, Int>()
        for (i in 0..<nums.size) {
            map[nums[i]]?.let {
                map[nums[i]] = (map[nums[i]] ?: 0) + 1
            } ?: run {
                map[nums[i]] = 0
                map[nums[i]] = (map[nums[i]] ?: 0) + 1
            }
        }

        return map.entries.sortedByDescending { it.value }.take(k).map { it.key }
        .toIntArray()
    }
}

File: Arrays_Hashing/TopKFrequentElements/TopKFrequentElements.swift
class Solution {
    func topKFrequent(_ nums: [Int], _ k: Int) -> [Int] {
        var dict: [Int: Int] = [:]
        
        for n in nums {
            if dict[n] == nil {
                dict[n] = 1
            } else {
                dict[n]! += 1
            }
        }

        return dict.sorted { $0.value > $1.value }
           .prefix(k)
           .map { $0.key }
    }
}

/*
dict[n]! += 1
// is sugar for
dict[n]! = dict[n]! + 1

*/
